docs/memdbg.man
.\" LIBDIR
.TH "MEMDBG" "1rko" "21 Feb 2002"
.SH NAME
memdbg \-  analyzes the MEMDEBUG output

.SH SYNOPSIS

 memdbg \fR[\fB-h\fR|\fB-m\fR|\fB-o output\fR][\fB--\fR] [input \.\.\.]

.SH DESCRIPTION
.I memdbg
\- analyzes the MEMDEBUG output checking that allocations have
associated free-s.

.TP 15
-h
this helpful message
.TP
-m
output the MEMDEBUG lines, else the default is to suppress them
and only output the regular input lines.
.TP
-o output
The output file name, else default to
.IR stdout .
.TP
--
stop option parsing (useful for filenames that start with `-')
.TP
input \.\.\.
input file names to read, else default to
.I stdin
if not specified.
.TP
exit status
is set to the number of memory discrepancies found.  Hence if the exit
status is 0 then there probably where no errors, but the output report
should be viewed anyways.

.SH EXAMPLES
The bulk of the task is performed by the librko memdebug
facility to generate the MEMDEBUG output
.P
Suppose the following was generated by MEMDEBUG (see
the example in memdebug(3rko)).
.nf
Some regular output
RKOMEM:         malloc : tmemdebug.c 36 : 0x8049fd8 80
RKOMEM:         free   : tmemdebug.c 38 : 0x8049fd8
RKOMEM:         malloc : tmemdebug.c 41 : 0x8049fd8 160
RKOMEM: realloc_free   : tmemdebug.c 43 : 0x8049fd8
RKOMEM: realloc_malloc : tmemdebug.c 43 : 0x8049fd8 320
RKOMEM:         malloc : tmemdebug.c 46 : 0x804a120 160
RKOMEM:         calloc : tmemdebug.c 49 : 0x804a1c8 20 24
RKOMEM:         free   : tmemdebug.c 51 : 0x804a1c8
RKOMEM:         calloc : tmemdebug.c 54 : 0x804a1c8 40 24
RKOMEM: realloc_free   : tmemdebug.c 57 : 0x804a120
RKOMEM: realloc_malloc : tmemdebug.c 57 : 0x804a590 320
RKOMEM:         free   : tmemdebug.c 59 : 0x8049fd8
RKOMEM:         malloc : tmemdebug.c 62 : 0x8049fd8 20
Some more regular output
RKOMEM:         free   : tmemdebug.c 67 : 0x804a590
RKOMEM:         free   : tmemdebug.c 68 : 0x804a1c8
RKOMEM:         free   : tmemdebug.c 69 : 0x8049fd8
RKOMEM:         free   : tmemdebug.c 70 : (null)
.fi
The result after passing the above text through
.I memdbg
will like this:
.nf

Some regular output
Some more regular output
        malloc count = 4
        calloc count = 2
realloc_malloc count = 2
realloc_free   count = 2
        free   count = 6
	>>> unbalanced alloc/free count

Questionable memory calls:
	>>> superfluous free at 'tmemdebug.c 70' of '(null)'

.fi
The exit status was 0 since
.I memdbg
does not consider a superfluous free to be an error.

.SH SEE ALSO
memdebug(3rko), calloc(3), free(3), malloc(3), realloc(3)


.SH AUTHOR
R.K.Owen,Ph.D.

.KEY WORDS


NEWS

librko-1.0.0	2002/10/21
* first release with a working configure, using automake/autoconf etc.
* added spawn3



memdbg.c
static const char RCSID[]="@(#)$Id: memdbg.c,v 1.5 2002/09/13 01:50:44 rk Exp $";
static const char AUTHOR[]="@(#)memdbg 1.0 2002/02/16 R.K.Owen,PhD";

/* memdbg - parses through the output generated by memdebug
 * 	analyzing the memory allocation/free calls.
 * 	Memdebug is not a replacement library for the memory
 * 	calls ... but a macro substitution for them.
 *
 *	Supports only the ANSI-C routines:
 *		  calloc,   free,   malloc,   realloc
 *	that have been substituted with the following names:
 *		m_calloc, m_free, m_malloc, m_realloc
 *	the output goes to stderr.
 */
/** ** Copyright *********************************************************** **
 ** 									     **
 ** Copyright 2002 by R.K.Owen,Ph.D.		                      	     **
 ** last known email: librko@kooz.sj.ca.us				     **
 **                   rk@owen.sj.ca.us					     **
 ** see LICENSE.LGPL, which must be provided, for details		     **
 ** 									     **
 ** ************************************************************************ **/

#include <stdio.h>
#include <string.h>
#include <unistd.h>	/* getopt */
#include "avec.h"
#include "uvec.h"
#ifdef RKOERROR
#  include "rkoerror.h"
#else
#  define rkoperror(x) {}
#endif

/* global variables */
int out_mem = 0;
avec *calls;
avec *memlocs;
uvec *memlog;
char wbuf[257];

char *memcalls[] = {
	"malloc",
	"calloc",
	"realloc",
	"free",
	(char *) NULL
};

/* convert a void *  to a long */
long v2l (void *vptr) {
	unsigned int	icount;
	unsigned short	scount;
	unsigned long	lcount;

	if (sizeof(void *) == sizeof(unsigned int)) {
		icount = (unsigned int) vptr;
		lcount = icount;
	} else if (sizeof(void *) == sizeof(unsigned short)) {
		scount = (unsigned short) vptr;
		lcount = scount;
	} else if (sizeof(void *) == sizeof(unsigned long)) {
		lcount = (unsigned long) vptr;
	} else {
		lcount = (unsigned long) vptr;
	}
	return lcount;
}

int initial_memdbg (void) {
	if (avec_set_fns(AVEC_COUNT,NULL)) return -2;
	if (!(calls = avec_ctor(10))) return -1;

	if (avec_set_fns(AVEC_STDC,NULL)) return -4;
	if (!(memlocs = avec_ctor(50))) return -3;

	if (avec_resize_percentage(memlocs,40)) return -5;

	if (!(memlog = uvec_ctor(20))) return -6;

	return 0;
}

int final_memdbg (FILE *output) {
	int retval = 0;
	avec_element **waeptr = (avec_element **) NULL;

	unsigned long	malloc_count = 0, calloc_count = 0,
			free_count = 0,
			realloc_malloc_count = 0, realloc_free_count = 0;
	int flag = 0;
	char **mlog;

	malloc_count = v2l(avec_lookup(calls, "malloc"));
	calloc_count = v2l(avec_lookup(calls, "calloc"));
	realloc_malloc_count = v2l(avec_lookup(calls, "realloc_malloc"));
	realloc_free_count   = v2l(avec_lookup(calls, "realloc_free"));
	free_count = v2l(avec_lookup(calls, "free"));

	fprintf(output, "        malloc count = %ld\n",  malloc_count);
	fprintf(output, "        calloc count = %ld\n",  calloc_count);
	fprintf(output, "realloc_malloc count = %ld\n",  realloc_malloc_count);
	fprintf(output, "realloc_free   count = %ld\n",  realloc_free_count);
	fprintf(output, "        free   count = %ld\n",  free_count);

	retval = malloc_count + calloc_count + realloc_malloc_count
		- free_count - realloc_free_count;

	if (retval) {
		fprintf(output, "\t>>> unbalanced alloc/free count\n");
	}
	/* abs(retval) */
	retval = (retval > 0 ? retval : -retval);

	while ((waeptr = avec_walk_r(memlocs, waeptr))) {
		if (!flag) {
			fprintf(output,"\nDangling memory allocations:\n");
			flag++;
		}
		fprintf(output,"\t>>> %s : %s\n",(*waeptr)->key,
			(char *) (*waeptr)->data);
		retval++;
	}

	if (uvec_number(memlog)) {
		fprintf(output,"\nQuestionable memory calls:\n");
		mlog = (char **) uvec_vector(memlog);
		while (*mlog) {
			fprintf(output,"\t>>> %s\n",*mlog);
			mlog++;
		}
	}
	/* clean up memory */
	uvec_dtor(&memlog);
	avec_dtor(&memlocs);
	avec_dtor(&calls);

	return retval;
}

int process_mem(FILE *output,
	char const *mcall, char const *src, char const *mloc) {

	char *lval;
	char mbuf[512];

	lval = (char *) avec_lookup(memlocs, mloc);

#ifdef RKOERROR
	if (!lval && rkoerrno != RKO_OK) rkoperror("memdbg : process_mem");
#endif

	if (lval && *lval) { /* found entry */
		if (!strcmp("malloc",mcall)) {
			strcpy(mbuf,"malloc overwrite of '");
			strncat(mbuf,lval, 100);
			strcat(mbuf,"' at '");
			strncat(mbuf,src, 100);
			strcat(mbuf,"' for '");
			strncat(mbuf,mloc, 100);
			strcat(mbuf,"'");
			if (out_mem) fprintf(output,">>> %s\n", mbuf);
			if (uvec_add(memlog, mbuf) < 0)
				rkoperror("memdbg : process_mem");
		} else if (!strcmp("free",mcall)) {
			/* this is normal */
			if (avec_delete(memlocs, mloc) < 0)
				rkoperror("memdbg : process_mem");
			if (avec_insert(calls,mcall) < 0)
				rkoperror("memdbg : process_mem");
		} else if (!strcmp("calloc",mcall)) {
			strcpy(mbuf,"calloc overwrite of '");
			strncat(mbuf,lval, 100);
			strcat(mbuf,"' at '");
			strncat(mbuf,src, 100);
			strcat(mbuf,"' for '");
			strncat(mbuf,mloc, 100);
			strcat(mbuf,"'");
			if (out_mem) fprintf(output,">>> %s\n", mbuf);
			if (uvec_add(memlog, mbuf) < 0)
				rkoperror("memdbg : process_mem");
		} else if (!strcmp("realloc_free",mcall)) {
			/* this is normal */
			if (avec_delete(memlocs, mloc) < 0)
				rkoperror("memdbg : process_mem");
			if (avec_insert(calls,mcall) < 0)
				rkoperror("memdbg : process_mem");
		} else if (!strcmp("realloc_malloc",mcall)) {
			/* this is normal */
			strncpy(mbuf,mcall,30);
			strcat(mbuf," ");
			strncat(mbuf,src,100);
			if (avec_insert(memlocs, mloc, mbuf) < 0)
				rkoperror("memdbg : process_mem");
			if (avec_insert(calls,mcall) < 0)
				rkoperror("memdbg : process_mem");
		} else {
			if (out_mem) fprintf(output,
			">>> unknown call '%s' at '%s'\n", mcall, mloc);
		}
	} else {	/* no entry */
		if (!strcmp("malloc",mcall)) {
			/* this is normal */
			strncpy(mbuf,mcall,30);
			strcat(mbuf," ");
			strncat(mbuf,src,100);
			if (avec_insert(memlocs, mloc, mbuf) < 0)
				rkoperror("memdbg : process_mem");
			if (avec_insert(calls,mcall) < 0)
				rkoperror("memdbg : process_mem");
		} else if (!strcmp("free",mcall)) {
			if (!strcmp("(null)",mloc)) {
				strcpy(mbuf,"superfluous free at '");
				strncat(mbuf,src, 100);
				strcat(mbuf,"' of '");
				strncat(mbuf,mloc, 100);
				strcat(mbuf,"'");
				if (out_mem) fprintf(output,">>>%s\n", mbuf);
				if (uvec_add(memlog, mbuf) < 0)
					rkoperror("memdbg : process_mem");
			} else {
				if (avec_delete(memlocs, mloc) < 0)
					rkoperror("memdbg : process_mem");
				if (avec_insert(calls,mcall) < 0)
					rkoperror("memdbg : process_mem");
			}
		} else if (!strcmp("calloc",mcall)) {
			/* this is normal */
			strncpy(mbuf,mcall,30);
			strcat(mbuf," ");
			strncat(mbuf,src,100);
			if (avec_insert(memlocs, mloc, mbuf) < 0)
				rkoperror("memdbg : process_mem");
			if (avec_insert(calls,mcall) < 0)
				rkoperror("memdbg : process_mem");
		} else if (!strcmp("realloc_free",mcall)) {
			/* this is normal */
			if (strcmp("(null)",mloc)) {
				if (avec_delete(memlocs, mloc) < 0)
					rkoperror("memdbg : process_mem");
				if (avec_insert(calls,mcall) < 0)
					rkoperror("memdbg : process_mem");
			}
		} else if (!strcmp("realloc_malloc",mcall)) {
			strncpy(mbuf,mcall,30);
			strcat(mbuf," ");
			strncat(mbuf,src,100);
			if (avec_insert(memlocs, mloc, mbuf) < 0)
				rkoperror("memdbg : process_mem");
			if (avec_insert(calls,mcall) < 0)
				rkoperror("memdbg : process_mem");
		} else {
			if (out_mem) fprintf(output,
			">>> unknown call '%s' at '%s'\n", mcall, mloc);
		}
	}
	return 0;
}

int process_line(FILE *output, char *line) {
	char *ptr;
	char mcall[30], src[128], mloc[64];

	if ((ptr = strstr(line, "RKOMEM:"))) {
		/* found a MEMDEBUG message */
		/* print everything upto the this point */
		if (ptr != line) {
			*ptr = '\0';
			fprintf(output,"%s",line);
		}
		if (out_mem) {
			fprintf(output,"R%s",ptr + 1);
		}
		ptr += 7;
		if (3 != sscanf(ptr,"%30s : %128[^:] : %64s",mcall,src,mloc))
			return -1;
		/* set last ' ' to '\0' */
		if (src[strlen(src)-1] == ' ') src[strlen(src)-1] = '\0';
		(void) process_mem(output, mcall, src, mloc);
	} else {
		/* just a regular line */
		fprintf(output,"%s",line);
	}
	return 0;
}

int process_input(FILE *input, FILE *output) {
	char *wptr = wbuf, *ptr;
	int wbufsize = sizeof(wbuf)/sizeof(*wbuf);
	size_t wlen = 0;
	size_t numread;

	/* make sure wbuf ends with '\0' */
	wbuf[--wbufsize] = '\0';

	/* read into buffer */
	while ((numread = fread(wptr,1,wbufsize - wlen,input) + wlen)) {
		/* set all \n to '\0' */
		ptr = wptr;			/* do only what was read */
		while (ptr - wptr < numread) {
			if (*ptr == '\n') *ptr = '\0';
			ptr++;
		}

		/* search for strings */
		ptr = wptr = wbuf;			/* look at full wbuf */
		while (ptr - wbuf < numread) {
			if (!*ptr) {	/* each '\0' is a '\n' */
				fprintf(output,"\n");
				ptr++;
			       	continue;
			}

			wlen = strlen(ptr);		/* length of line */

			if (ptr - wbuf + wlen >= wbufsize) {
			/* at end of block and possibly a split line */
				strcpy(wbuf, ptr);
				wptr = wbuf + wlen;
				ptr = wptr + numread;	/* end of block */
			} else {
				process_line(output, ptr);
				if ((! out_mem) && !strncmp(ptr,"RKOMEM:",7))
					ptr++;		/* skip '\0' */
				ptr += wlen;
				wlen = 0;
			}
		}
	}
	return 0;
}

void usage(char *ProgName) {

	(void) printf(
"\n"
"%s - analyzes the MEMDEBUG output checking that allocations have\n"
"	associated free-s (StdC only)\n"
"\n"
"usage: %s [-h][-m][-o output][--][input ...]\n"
"	-h		this helpful information\n"
"	-m		output the MEMDEBUG lines (default = only summary)\n"
"	-o output	output file name (else defaults to stdout)\n"
"	--		stops further option processing\n"
"	input ...	one or more input file names (else defaults to stdin)\n"
"			the input of '-' will use stdin at that place\n"
"\n"
"output: the summary prints any non-match memory allocations, usually\n"
"	due to a memory leak.\n\n",
		ProgName, ProgName);
}

int main(int argc, char *argv[]) {
	int retval = 0;
	char *PrgNm = argv[0];	/* must send entire argv to getopt */
	FILE *input, *output;
	int i;

	input = stdin;
	output = stdout;

	if ((retval = initial_memdbg())) {
		fprintf(stderr,"%s : initial_memdbg failure = '%d'\n",
			PrgNm, retval);
		return retval;
	}

	while ((i = getopt(argc,argv,"hmo:")) != -1) {
		switch (i) {
		case 'h':
			usage(PrgNm);
			return 0;
			break;
		case 'm':
			out_mem = 1;
			break;
		case 'o':
			if (!(output = fopen(optarg, "w"))) {
				fprintf(stderr,"Can't open %s for writing!\n\n",
					optarg);
				usage(PrgNm);
				return 2;
			}
			break;
		case '?':
		default:
			usage(PrgNm);
			return 1;
			break;
		}
	}

	/* cycle through arguments */
	argv = &(argv[optind]);
	if (*argv) {
		while (*argv) {				/* step through args */
			if (!strcmp(*argv,"-")) {	/* want stdin */
				input = stdin;
			} else {
				if(!(input = fopen(*argv,"r"))) {
					fprintf(stderr,
					"\nCan't open %s for reading!\n\n",
						*argv);
				}
			}
			argv++;
			retval = process_input(input,output);
			fclose(input);
		}
	} else {					/* no arg hence stdin */
		retval = process_input(input,output);
	}

	retval = final_memdbg(output);
	fclose(output);

	return retval;
}
